File: frontend/types/websocket.ts
export type AssistantState =
  | 'idle'
  | 'listening'
  | 'transcribing'
  | 'thinking'
  | 'executing_tool'
  | 'responding'
  | 'speaking'
  | 'interrupted'
  | 'error';

export type WSMessageType =
  // Client -> Server
  | 'user_message'
  | 'start_listening'
  | 'stop_speaking'
  | 'cancel_request'
  | 'new_conversation'
  | 'ping'
  // Server -> Client
  | 'state_change'
  | 'transcription_started'
  | 'transcription_completed'
  | 'thinking_started'
  | 'tool_execution_started'
  | 'tool_execution_completed'
  | 'response_chunk'
  | 'response_complete'
  | 'request_cancelled'
  | 'error'
  | 'pong'
  | 'conversation_created';

export interface WSClientMessage {
  type: WSMessageType;
  request_id: string;
  conversation_id?: number;
  timestamp: string;
  data: Record<string, unknown>;
}

export interface WSServerMessage {
  type: WSMessageType;
  request_id: string;
  conversation_id?: number;
  timestamp: string;
  data: Record<string, unknown>;
}

// Client message types
export interface UserMessage extends WSClientMessage {
  type: 'user_message';
  data: { content: string };
}

export interface StartListeningMessage extends WSClientMessage {
  type: 'start_listening';
}

export interface StopSpeakingMessage extends WSClientMessage {
  type: 'stop_speaking';
}

export interface CancelRequestMessage extends WSClientMessage {
  type: 'cancel_request';
}

export interface NewConversationMessage extends WSClientMessage {
  type: 'new_conversation';
}

export interface PingMessage extends WSClientMessage {
  type: 'ping';
}

// Server message types
export interface StateChangeMessage extends WSServerMessage {
  type: 'state_change';
  data: { state: AssistantState };
}

export interface TranscriptionStartedMessage extends WSServerMessage {
  type: 'transcription_started';
}

export interface TranscriptionCompletedMessage extends WSServerMessage {
  type: 'transcription_completed';
  data: { text: string; is_final: boolean };
}

export interface ThinkingStartedMessage extends WSServerMessage {
  type: 'thinking_started';
}

export interface ToolExecutionStartedMessage extends WSServerMessage {
  type: 'tool_execution_started';
  data: { name: string; args: Record<string, unknown> };
}

export interface ToolExecutionCompletedMessage extends WSServerMessage {
  type: 'tool_execution_completed';
  data: { name: string; result: unknown; error: string | null };
}

export interface ResponseChunkMessage extends WSServerMessage {
  type: 'response_chunk';
  data: { content: string };
}

export interface ResponseCompleteMessage extends WSServerMessage {
  type: 'response_complete';
}

export interface RequestCancelledMessage extends WSServerMessage {
  type: 'request_cancelled';
  data: { request_id: string };
}

export interface ErrorMessage extends WSServerMessage {
  type: 'error';
  data: { message: string; code: string };
}

export interface PongMessage extends WSServerMessage {
  type: 'pong';
}

export interface ConversationCreatedMessage extends WSServerMessage {
  type: 'conversation_created';
  data: { conversation_id: number };
}

export type AnyClientMessage =
  | UserMessage
  | StartListeningMessage
  | StopSpeakingMessage
  | CancelRequestMessage
  | NewConversationMessage
  | PingMessage;

export type AnyServerMessage =
  | StateChangeMessage
  | TranscriptionStartedMessage
  | TranscriptionCompletedMessage
  | ThinkingStartedMessage
  | ToolExecutionStartedMessage
  | ToolExecutionCompletedMessage
  | ResponseChunkMessage
  | ResponseCompleteMessage
  | RequestCancelledMessage
  | ErrorMessage
  | PongMessage
  | ConversationCreatedMessage;

File: backend/app/schemas/websocket.py
from datetime import datetime
from enum import Enum
from typing import Any
from uuid import UUID

from pydantic import BaseModel, Field


class AssistantState(str, Enum):
    IDLE = "idle"
    LISTENING = "listening"
    TRANSCRIBING = "transcribing"
    THINKING = "thinking"
    EXECUTING_TOOL = "executing_tool"
    RESPONDING = "responding"
    SPEAKING = "speaking"
    INTERRUPTED = "interrupted"
    ERROR = "error"


class WSMessageType(str, Enum):
    # Client -> Server
    USER_MESSAGE = "user_message"
    START_LISTENING = "start_listening"
    STOP_SPEAKING = "stop_speaking"
    CANCEL_REQUEST = "cancel_request"
    NEW_CONVERSATION = "new_conversation"
    PING = "ping"

    # Server -> Client
    STATE_CHANGE = "state_change"
    TRANSCRIPTION_STARTED = "transcription_started"
    TRANSCRIPTION_COMPLETED = "transcription_completed"
    THINKING_STARTED = "thinking_started"
    TOOL_EXECUTION_STARTED = "tool_execution_started"
    TOOL_EXECUTION_COMPLETED = "tool_execution_completed"
    RESPONSE_CHUNK = "response_chunk"
    RESPONSE_COMPLETE = "response_complete"
    REQUEST_CANCELLED = "request_cancelled"
    ERROR = "error"
    PONG = "pong"
    CONVERSATION_CREATED = "conversation_created"


class WSClientMessage(BaseModel):
    type: WSMessageType
    request_id: str = Field(default_factory=lambda: str(UUID(int=0)))  # Will be overridden
    conversation_id: int | None = None
    timestamp: datetime = Field(default_factory=datetime.utcnow)
    data: dict = Field(default_factory=dict)


class WSServerMessage(BaseModel):
    type: WSMessageType
    request_id: str
    conversation_id: int | None = None
    timestamp: datetime = Field(default_factory=datetime.utcnow)
    data: dict = Field(default_factory=dict)


# Client message types
class UserMessage(WSClientMessage):
    type: WSMessageType = WSMessageType.USER_MESSAGE
    data: dict = Field(default_factory=lambda: {"content": ""})


class StartListeningMessage(WSClientMessage):
    type: WSMessageType = WSMessageType.START_LISTENING


class StopSpeakingMessage(WSClientMessage):
    type: WSMessageType = WSMessageType.STOP_SPEAKING


class CancelRequestMessage(WSClientMessage):
    type: WSMessageType = WSMessageType.CANCEL_REQUEST


class NewConversationMessage(WSClientMessage):
    type: WSMessageType = WSMessageType.NEW_CONVERSATION


class PingMessage(WSClientMessage):
    type: WSMessageType = WSMessageType.PING


# Server message types
class StateChangeMessage(WSServerMessage):
    type: WSMessageType = WSMessageType.STATE_CHANGE
    data: dict = Field(default_factory=lambda: {"state": "idle"})


class TranscriptionStartedMessage(WSServerMessage):
    type: WSMessageType = WSMessageType.TRANSCRIPTION_STARTED


class TranscriptionCompletedMessage(WSServerMessage):
    type: WSMessageType = WSMessageType.TRANSCRIPTION_COMPLETED
    data: dict = Field(default_factory=lambda: {"text": "", "is_final": True})


class ThinkingStartedMessage(WSServerMessage):
    type: WSMessageType = WSMessageType.THINKING_STARTED


class ToolExecutionStartedMessage(WSServerMessage):
    type: WSMessageType = WSMessageType.TOOL_EXECUTION_STARTED
    data: dict = Field(default_factory=lambda: {"name": "", "args": {}})


class ToolExecutionCompletedMessage(WSServerMessage):
    type: WSMessageType = WSMessageType.TOOL_EXECUTION_COMPLETED
    data: dict = Field(default_factory=lambda: {"name": "", "result": None, "error": None})


class ResponseChunkMessage(WSServerMessage):
    type: WSMessageType = WSMessageType.RESPONSE_CHUNK
    data: dict = Field(default_factory=lambda: {"content": ""})


class ResponseCompleteMessage(WSServerMessage):
    type: WSMessageType = WSMessageType.RESPONSE_COMPLETE


class RequestCancelledMessage(WSServerMessage):
    type: WSMessageType = WSMessageType.REQUEST_CANCELLED
    data: dict = Field(default_factory=lambda: {"request_id": ""})


class ErrorMessage(WSServerMessage):
    type: WSMessageType = WSMessageType.ERROR
    data: dict = Field(default_factory=lambda: {"message": "", "code": ""})


class PongMessage(WSServerMessage):
    type: WSMessageType = WSMessageType.PONG


class ConversationCreatedMessage(WSServerMessage):
    type: WSMessageType = WSMessageType.CONVERSATION_CREATED
    data: dict = Field(default_factory=lambda: {"conversation_id": 0})


# Union types for validation
AnyClientMessage = (
    UserMessage
    | NewConversationMessage
    | CancelRequestMessage
    | PingMessage
)

AnyServerMessage = (
    StateChangeMessage
    | TranscriptionStartedMessage
    | TranscriptionCompletedMessage
    | ThinkingStartedMessage
    | ToolExecutionStartedMessage
    | ToolExecutionCompletedMessage
    | ResponseChunkMessage
    | ResponseCompleteMessage
    | RequestCancelledMessage
    | ErrorMessage
    | PongMessage
    | ConversationCreatedMessage
)